compiler/src/dotty/tools/backend/jvm/BCodeSkelBuilder.scala

```diff
@@ -107,14 +110,92 @@ trait BCodeSkelBuilder extends BCodeHelpers {
 
       initJClass(cnode)
 
-      val methodSymbols = for (f <- cd.symbol.info.decls.toList if f.is(Method) && f.isTerm && !f.is(Module)) yield f
-      val hasStaticCtor = methodSymbols exists (_.isStaticConstructor)
-      if (!hasStaticCtor) {
-        // but needs one ...
-        if (isCZStaticModule || isCZParcelable) {
-          fabricateStaticInit()
+      val cd = if (isCZStaticModule) {
+        // Move statements from the primary constructor following the superclass constructor call to
+        // a newly synthesised tree representing the "<clinit>", which also assigns the MODULE$ field.
+        // Because the assigments to both the module instance fields, and the fields of the module itself
+        // are in the <clinit>, these fields can be static + final.
+
+        // Should we do this transformation earlier, say in Constructors? Or would that just cause
+        // pain for scala-{js, native}?
+        //
+        // @sjrd (https://github.com/lampepfl/dotty/pull/9181#discussion_r457458205):
+        // moving that before the back-end would make things significantly more complicated for
+        // Scala.js and Native. Both have a first-class concept of ModuleClass, and encode the
+        // singleton pattern of MODULE$ in a completely different way. In the Scala.js IR, there
+        // even isn't anything that corresponds to MODULE$ per se.
+        //
+        // So if you move this before the back-end, then Scala.js and Scala Native will have to
+        // reverse all the effects of this transformation, which would be counter-productive.
+
+
+        // TODO: enable once we change lazy val encoding
```

Comment:
To not forget about this, I suggest explaining this issue in https://github.com/lampepfl/dotty/issues/7140 and leaving a link to it here (also I'm surprised that this doesn't cause other issues on Java 9)i